rfcs/text/0000_alert_actions_service.md

```diff
@@ -0,0 +1,210 @@
+below for details how to create them
+
+## Creating Instances
+
+Instances are saved in the Kibana index and are used in order to `fire` an action with parameters. The
+`params` required are specific to the `actionType`. Changing the `actionType` also
+changes what `connectorTypes` are available for it.
+
+
+```JS
+server.actions.instance({
+  name: 'send message to slack',
+  actionType: 'notification',
+  connectorType: 'slack',
+  params: {
+    destination: '<slack url>',
+  },
+  connectorParams: {
+    channel: '#bot-playground',
+  },
+});
+```
+
+*Note: instances are saved objects that are space aware*
+
+## The Connectors
+
+The `connector` is what defines what happens when an action is *fired*. This
+will define the particular integration such as slack, email, etc.
+
+```JS
+server.actions.registerConnector({
+  actionType: 'notification',
+  connectorType: 'slack',
+  async handler({ params, connectorParams }) {
+    try {
+      const body = JSON.stringify({
+        ...connectorParams,
+        username: 'webhookbot',
+        text: params.message,
+        icon_emoji: ':ghost:',
+      });
+      const result = new url.URL(params.destination);
+      await fetch(result.toString(), {
+        method: 'POST',
+        body,
+      });
+    } catch (e) {
+      warn(`[slack] failed with ${e.message}`);
+    }
+  },
+});
+```
+
+## Defining Action Types
+
+Registering action types is the way that these actions are exposed to the
+alerts. Included in the definition are the required parameters that the alert
+should provide and those that are handed off to the *connectors* to fit into
+whatever service integration they provide.
+
+```JS
+server.actions.registerActionType({
+  name: 'notification',
+  fire: [
+    { name: 'destination', type: 'string' },
+    { name: 'message', type: 'string' },
+    { name: 'title', type: 'string', optional: true },
+  ],
+});
+```
+
+## Retrieve a list of available actions
+
+If you want to see the available list of actions that are registered you can get
+that list from the `server.actions` service.
+
+```JS
+server.actions.available(); // returns ['send message to slack']
+```
+
+Or from the front-end there is a REST API that will return the available actions.
+
+```sh
+$ curl -X GET localhost:5601/api/actions
+```
+
+# Drawbacks
+
+The relationship between **action types** and **connectors** is a bit complex as
+well as the supported parameter meta types.
+
+Another downside to this design is that an alert must be changed in order to support
+**action types**. So as new action types are added to meet users needs are
+discovered alerts will require changes to support the new types. This increases
+the code complexity of each alert and time to market for new integrations. And
+increased risk of alerts for specific applications not having options to perform
+the same actions that other applications support. Also differences in UI and
+customizability for the action connectors may vary per application.
```

Comment:
> So as new action types are added to meet users needs are discovered alerts will require changes to support the new types.

Can you elaborate on this a bit? I assumed an alert could be configured by a user to use specific action configurations (like send notification to slack). If we add a new action handler for the "notification" actionType, I don't see why we'd need to change the alerting code to be able to use it. Of course, the user would still to configure them, but I assumed that by adding a new action handler, it would show up in a list of options for the user to configure.

Am I understanding this incorrectly?

Reply:
Yeah let me try to elaborate and may be it isn't as big of a concern.

One of the main methods this service will be consumed is by the alert tasks which will specifically invoke a method to use an action type something like the below.

```JS
task({ actions }) {
  elasticsearch.search(someQuery...)
  ...
  actions.sendNotification({
      destination: "slack/url",
      message: params.message,
      title: params.title,
  })
}
```

The alert could also be coded to support sending one notification for each hit from elasticsearch or each that met some condition. What is possible with a given alert is up the alert itself. 

So given that the alert implementation must know the action type parameters. If we add other action types in the future all alert code would need to be updated to support those action types,  we wouldn't automatically be able to use the new types everywhere.

Reply:
Ok I got it. So if new **handlers** are added, the alerting code doesn't need to be updated. A user would be able to configure an **instance** of the action type that uses the new handler. This works because the **actionType** has consistent params for all handlers of that type.

When new **actionTypes** are added, alerting code would need to be updated because that type may accept completely different params. Makes sense to me now.